feature(KEP-4832): asynchronous preemption

sanposhiho opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind feature

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

The core implementation for KEP-4832, moving the part making API calls to goroutine inside the preemption.

I confirmed that this feature:
- makes the latency of PostFilter 1/5 
  - This is when each preemption preempts 3 Pods. The more Pods to kill, the bigger the improvement is, vice versa.
- improves the throughput by double in the scenario of `PreemptionAsync` with multiple runs of scheduler-perf.

See: https://github.com/kubernetes/kubernetes/pull/128170#issuecomment-2439369996

---

The note on the implementation detail:

**About Activate()**

The preemption plugin moves the preemptor pod to activeQ when the goroutine fails.

This is because if, for example, the deletion API fails and no Pod/Delete happens, 
the preemptor Pod could get stuck at the unsched pod pool. (since no event moves the Pod)

For this operation, we reuse the existing `Activate()` method on the queue that moves the Pod from unsched to activeQ.
And, to address the case where the goroutine fails immediately and the preemptor pod hasn't come back to the queue yet, we utilize in-flight event and `moveRequestCycle` like we do with cluster events.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #128020
Fixes https://github.com/kubernetes/kubernetes/issues/128019

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Added alpha support for asynchronous Pod preemption.
When the `SchedulerAsyncPreemption` feature gate is enabled, the scheduler now runs API calls to trigger preemptions asynchronously for better performance.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:

- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]:  http://kep.k8s.io/4832
```

sanposhiho committed:
feature(KEP-4832): asynchronous preemption

sanposhiho commented:
/test all

This PR is still WIP as I need to add more tests.

sanposhiho commented:
/hold

need to go thru the approver's approval.

AxeZhan commented on pkg/scheduler/framework/preemption/preemption.go:
```diff
@@ -401,6 +452,71 @@ func (ev *Evaluator) prepareCandidate(ctx context.Context, c Candidate, pod *v1.
+		// nomination updates these pods and moves them to the active queue. It
+		// lets scheduler find another place for them.
+		nominatedPods := getLowerPriorityNominatedPods(logger, ev.Handler, pod, c.Name())
+		if err := util.ClearNominatedNodeName(ctx, ev.Handler.ClientSet(), nominatedPods...); err != nil {
```
We're not clearing this in `prepareCandidate`(when feature is disabled) ?

sanposhiho replied:
Currently, we are. 
But, as mentioned in https://github.com/kubernetes/kubernetes/pull/128067, I have to think a bit more about it through the weekend.

sanposhiho replied:
We should do before.
Because after the moment the high priority pod gets nominated to the node, lower priority pods could be unschedulable, regardless of whether the API calls are completely done or not.

sanposhiho replied:
But, I still believe we don't need to clear nominated node names of all lower priority pods on the node.
Because, for the extreme example, if high priority pod only requires 2 CPU, and there are 10 lower priority Pods that require 2 CPU each and have already got nominated to the node, it doesn't make sense to remove the nomination from all.

So, I do believe we should seek other way like:
https://github.com/kubernetes/kubernetes/pull/128067

But, I'm considering such enhancement separated from this KEP (i.e., no need to implement it together), and thus I wouldn't like to include the change here.
I was thinking I'd pursue when I've got time after Kubecon. (... or you can take over if you want! :))
Also, either way, I believe that kind of scenario (there's a lower priority pod that just completed the preemption on the same node) is fairly rare, and hence we needn't actively prioritize it.

AxeZhan replied:
I'm +1 for separate it from this KEP, it will be easily to roll back if sth goes wrong after the implementation.

> ... or you can take over if you want! :

I’ll look into this when I have some free time, but I haven't thought of a good solution yet also :(

sanposhiho replied:
The rough idea that I have is like:
- When calculating the victims on the node, we assume the nominated Pods are on the node already.
  - Lower priority nominated Pods could be selected as victims, while higher ones shouldn't.
- Then, if lower nominated Pods are selected, `prepareCandidate` (or `prepareCandidateAsync`) unnominates them instead of deleting them.

So, in short, the nominated info on lower pods will be un-nominated only when necessary.

sanposhiho replied:
But, apart from that idea, I'd like to still seek the simplest option like https://github.com/kubernetes/kubernetes/pull/128067. 
We have to think thoroughly if there's any corner case though, maybe we can just do that.

nayihz commented on pkg/scheduler/framework/preemption/preemption.go:
```diff
@@ -335,6 +374,45 @@ func (ev *Evaluator) SelectCandidate(ctx context.Context, candidates []Candidate
+				return err
+			}
+			logger.V(2).Info("Victim Pod is already deleted", "preemptor", klog.KObj(preemptor), "victim", klog.KObj(victim), "node", c.Name())
+			return nil
```
Maybe we don't need to return nil when pod has been deleted. Just delete this line?

sanposhiho replied:
But we shouldnt reach L408 and L411, right? Those log and event would be inappropriate in this case

sftim commented:
Changelog suggestion
```diff
-The scheduler now runs some expensive operations within the preemption asynchronously for better performance.
-The feature gate SchedulerAsyncPreemption gates this feature, which is disabled by default.
-
-See http://kep.k8s.io/4832 for more details.
+Added alpha support for asynchronous Pod preemption.
+When the `SchedulerAsyncPreemption` feature gate is enabled, the scheduler now runs some expensive operations asynchronously for better performance.
```

sanposhiho commented:
/test all

sanposhiho commented:
/test all

sanposhiho commented:
Yey, confirmed that this change makes the latency of PostFilter 1/5 and improves the throughput by double in the scenario of `PreemptionAsync` with multiple runs of scheduler-perf.

**master**

```
    {
      "data": {
        "Average": 39.2565,
        "Perc50": 22.16837820053101,
        "Perc90": 22.16837820053101,
        "Perc95": 22.16837820053101,
        "Perc99": 22.16837820053101
      },
      "unit": "ms",
      "labels": {
        "Metric": "scheduler_plugin_execution_duration_seconds",
        "Name": "BenchmarkPerfScheduling/PreemptionAsync/5000Nodes/namespace-3",
        "event": "not applicable",
        "extension_point": "PostFilter",
        "plugin": "DefaultPreemption",
        "result": "not applicable"
      }
    },
    {
      "data": {
        "Average": 479.1271557382307,
        "Perc50": 474.9948743303111,
        "Perc90": 788.9765942203558,
        "Perc95": 788.9765942203558,
        "Perc99": 788.9765942203558
      },
      "unit": "pods/s",
      "labels": {
        "Metric": "SchedulingThroughput",
        "Name": "BenchmarkPerfScheduling/PreemptionAsync/5000Nodes/namespace-3",
        "event": "not applicable",
        "extension_point": "not applicable",
        "plugin": "not applicable",
        "result": "not applicable"
      }
    },
```

**after this change**

```
    {
      "data": {
        "Average": 7.157875,
        "Perc50": 8.210510444641116,
        "Perc90": 9.524192115783695,
        "Perc95": 9.688402324676515,
        "Perc99": 9.819770491790774
      },
      "unit": "ms",
      "labels": {
        "Metric": "scheduler_plugin_execution_duration_seconds",
        "Name": "BenchmarkPerfScheduling/PreemptionAsync/5000Nodes(async-preemption)/namespace-3",
        "event": "not applicable",
        "extension_point": "PostFilter",
        "plugin": "DefaultPreemption",
        "result": "not applicable"
      }
    },
    {
      "data": {
        "Average": 835.7530034075595,
        "Perc50": 881.5466646277152,
        "Perc90": 964.1355815661577,
        "Perc95": 964.1355815661577,
        "Perc99": 964.1355815661577
      },
      "unit": "pods/s",
      "labels": {
        "Metric": "SchedulingThroughput",
        "Name": "BenchmarkPerfScheduling/PreemptionAsync/5000Nodes(async-preemption)/namespace-3",
        "event": "not applicable",
        "extension_point": "not applicable",
        "plugin": "not applicable",
        "result": "not applicable"
      }
    },
```

sanposhiho commented:
/test all

sanposhiho commented:
/cc @macsko @dom4ha @alculquicondor

sanposhiho commented on test/integration/scheduler/preemption/preemption_test.go:
Note for reviewers:

Basically, a new test is only `TestAsyncPreemption`. 
Other changes are just making all tests run both with/without the feature gate enabled.

sanposhiho commented:
/test all

sanposhiho commented on test/integration/scheduler/preemption/preemption_test.go:
Note for reviewers:

Basically, a new test is only `TestAsyncPreemption`. 
Other changes are just making all tests run both with/without the feature gate enabled.

sanposhiho commented:
/test all

sanposhiho commented on test/integration/scheduler/preemption/preemption_test.go:
Note for reviewers:

Basically, a new test is only `TestAsyncPreemption`. 
Other changes are just making all tests run both with/without the feature gate enabled.

sanposhiho commented:
/retest

macsko commented on test/integration/scheduler/preemption/preemption_test.go:
```diff
@@ -452,59 +463,561 @@ func TestPreemption(t *testing.T) {
+			},
+		},
+		{
+			name: "Higher priority Pod takes over the place for higher priority Pod that is running the preemption",
```
Maybe
```suggestion
			name: "Higher priority Pod takes over the place for lower priority Pod that is running the preemption",
```

macsko commented on test/integration/scheduler/preemption/preemption_test.go:
```diff
@@ -452,59 +463,561 @@ func TestPreemption(t *testing.T) {
+					if ch, ok := preemptionDoneChannels[preemptor.Name]; ok {
+						<-ch
+					}
+					err := preemptPodFn(ctx, c, preemptor, victim, pluginName)
+					if err != nil {
+						return err
+					}
+					return nil
```
```suggestion
					return preemptPodFn(ctx, c, preemptor, victim, pluginName)
```

macsko commented on test/integration/scheduler/preemption/preemption_test.go:
```diff
@@ -452,59 +463,561 @@ func TestPreemption(t *testing.T) {
+					if err := wait.PollUntilContextTimeout(testCtx.Ctx, time.Millisecond*200, wait.ForeverTestTimeout, false, func(ctx context.Context) (bool, error) {
+						activePods := testCtx.Scheduler.SchedulingQueue.PodsInActiveQ()
+						return len(activePods) != 0, nil
+					}); err != nil {
+						t.Fatalf("Expected the pod %s to be scheduled, but no pod arrives at the activeQ", scenario.schedulePod.podName)
 					}
-				} else {
-					if p.DeletionTimestamp != nil {
-						t.Errorf("Didn't expect pod %v to get preempted.", p.Name)
+
+					if err := wait.PollUntilContextTimeout(testCtx.Ctx, time.Millisecond*200, wait.ForeverTestTimeout, false, func(ctx context.Context) (bool, error) {
+						return testCtx.Scheduler.SchedulingQueue.PodsInActiveQ()[0].Name == scenario.schedulePod.podName, nil
+					}); err != nil {
+						t.Fatalf("The pod %s is expected to be scheduled, but the top Pod is %s", scenario.schedulePod.podName, testCtx.Scheduler.SchedulingQueue.PodsInActiveQ()[0].Name)
+					}
```
Do we need to poll twice here?

sanposhiho replied:
Good catch, I splitted one into two when debugging the problem. We can bring back them into one.

dom4ha commented:
/test pull-kubernetes-scheduler-perf

sanposhiho committed:
fix: address reviews

sanposhiho commented:
Addressed the reviews from @macsko, plus rebased for the conflict.

sanposhiho committed:
fix: fixed some tests

sanposhiho commented:
/retest

sanposhiho commented:
/test pull-kubernetes-scheduler-perf

sanposhiho committed:
fix: modify newly added preemption test

sanposhiho committed:
fix: refactor TestPrepareCandidate to test async feature too

sanposhiho committed:
feat: add the e2e test for the async preemption

sanposhiho commented:
I forgot adding the e2e tests although I mentioned it as the alpha requirement. Added the one.

macsko commented on test/e2e/scheduling/preemption.go:
```diff
@@ -317,6 +318,162 @@ var _ = SIGDescribe("SchedulerPreemption", framework.WithSerial(), func() {
+	*/
+	framework.It("validates various priority Pods preempt expectedly with the async preemption", feature.SchedulerAsyncPreemption, func(ctx context.Context) {
+		var podRes v1.ResourceList
+		// Create two pods per node that uses a lot of the node's resources.
```
```suggestion
		// Create 10 pods per node that uses a lot of the node's resources.
```

macsko commented on test/e2e/scheduling/preemption.go:
```diff
@@ -317,6 +318,162 @@ var _ = SIGDescribe("SchedulerPreemption", framework.WithSerial(), func() {
+
+			// Create 10 low priority pods on each node, which will use up 10/10 of the node's resources.
+			for j := 0; j < 10; j++ {
+				// Request 2 of the available resources for the victim pods
```
```suggestion
				// Request 1 of the available resources for the victim pods
```
?

macsko commented on test/e2e/scheduling/preemption.go:
```diff
@@ -317,6 +318,162 @@ var _ = SIGDescribe("SchedulerPreemption", framework.WithSerial(), func() {
+		ginkgo.By("Create 10 low-priority pods on each node.")
+		lowPriorityPods := make([]*v1.Pod, 0, 10*len(nodeList.Items))
+		// Create pods in the cluster.
+		// One of them has low priority, making it the victim for preemption.
```
I think this line is incorrect

macsko commented on test/e2e/scheduling/preemption.go:
```diff
@@ -317,6 +318,162 @@ var _ = SIGDescribe("SchedulerPreemption", framework.WithSerial(), func() {
+				if err != nil {
+					return false, err
+				}
+				return highPod.Status.NominatedNodeName != "", err
```
```suggestion
				return highPod.Status.NominatedNodeName != "", nil
```

sanposhiho commented on test/e2e/scheduling/preemption.go:
```diff
@@ -317,6 +318,162 @@ var _ = SIGDescribe("SchedulerPreemption", framework.WithSerial(), func() {
 	})
 
+	/*
+		Release: v1.23
```
🤦‍♂️

sanposhiho committed:
fix: address reviews on e2e

sanposhiho commented:
/retest

sanposhiho committed:
make sure DefaultPreemption implements PreEnqueuePlugin

sanposhiho committed:
feat: support metrics

sanposhiho commented:
/retest

sanposhiho commented:
@alculquicondor 
I've yet only got a few review path from other reviewers though, can you also take a look when you get a moment?
Since the code freeze is approaching as always, I'd like to prompt the speed as much as possible.

dom4ha reviewed (commented):
> @alculquicondor 
> I've yet only got a few review path from other reviewers though, can you also take a look when you get a moment?
> Since the code freeze is approaching as always, I'd like to prompt the speed as much as possible. 

I'm looking into it now

sanposhiho committed:
chore: tiny wording on the comment

AxeZhan commented:
> I've yet only got a few review path from other reviewers though, can you also take a look when you get a moment?
Since the code freeze is approaching as always, I'd like to prompt the speed as much as possible.

FYI, I've looked into the implementation and this part lgtm.
I'll take a look at test part later.

dom4ha commented on pkg/scheduler/framework/preemption/preemption.go:
```diff
@@ -203,17 +287,23 @@ func (ev *Evaluator) Preempt(ctx context.Context, pod *v1.Pod, m framework.NodeT
-	if status := ev.prepareCandidate(ctx, bestCandidate, pod, ev.PluginName); !status.IsSuccess() {
-		return nil, status
+	if ev.enableAsyncPreemption {
+		ev.prepareCandidateAsync(bestCandidate, pod, ev.PluginName)
```
In async version, preemption failures does not prevent scheduler from annotating pod as unschedulable and setting nominated node (in `handleSchedulingFailure`). Isn't there a risk that the pod will remain unschedulable forever?

sanposhiho replied:
Good point, I implemented `PodActivator` to deal with the failure scenario; we move the preemptor to activeQ when the preemption goroutine fails

dom4ha replied:
Not sure yet how the activator works, but I was thinking about something simpler. What about setting error status in `preempting` which would be read the next time the PreEnque is called?

Maybe I'm missing something obvious, but I've just noticed that the plugin does not register to any cluster events. I don't see us calling `moveToActiveQ` either, so how was it supposed to be unblocked?

sanposhiho replied:
> What about setting error status in preempting which would be read the next time the PreEnque is called?

The problem here is, for example, when the preemption API calls aren't successful at all, the preemptor Pod gets stuck in the unschedulable pod pool because no Pod/Delete event happens.
So, we have to somehow let the preemption plugin move the Pod out of the unsched pod pool. That's `Activate()`.

> Maybe I'm missing something obvious, but I've just noticed that the plugin does not register to any cluster events. I don't see us calling moveToActiveQ either, so how was it supposed to be unblocked?

We don't need to register any events in the preemption plugin.
For example, the pod is originally rejected by the node resource fit,
1. the preemptor pod gets `unschedulablePlugin: node-resource-fit`
2. the scheduling queue tries to requeue this pod every time node-resource-fit's QHint returns Queue, but the preemption plugin blocks the requeueing at PreEnqueue until all the preemption is done.
3. When the last pod is deleted, the scheduling queue, again, should tries to requeue the preemptor pod, and this requeueing goes through because the preemption plugin doesn't block this pod anymore at PreEnqueue.

dom4ha replied:
> We don't need to register any events in the preemption plugin because the preemptor Pod should have the other plugins in unschedulablePlugins.
> 
Isn't it something https://github.com/kubernetes/kubernetes/pull/127180 is going to change, so that only the GatedPlugin's events are supposed to be checked? Here we don't register any, but I think it's possible that preemption plugin will become the gated one. Note that `runPreEnqueuePlugins` can reassign to the preemption plugin, as it become a blocking one (for instance when resource-fit let the pod go for some other reason).

BTW. why we need `preempting` map then if we wait for 'resource-fit' plugin anyway?

dom4ha commented on pkg/scheduler/framework/preemption/preemption.go:
```diff
@@ -401,6 +461,85 @@ func (ev *Evaluator) prepareCandidate(ctx context.Context, c Candidate, pod *v1.
+			// We do not return as this error is not critical.
+		}
+
+		// We can evict all victims in parallel, but the last one.
+		// We have to remove the pod from the preempting map before the last one is evicted
+		// because, otherwise, the pod removal might be notified to the scheduling queue before
+		// we remove this pod from the preempting map,
+		// and the pod could end up stucking at the unschedulable pod pool
+		// by all the pod removal events being ignored.
+
```
IIUC this comment refers to Parallelizer and len -1, so you can move it below?

dom4ha commented on pkg/scheduler/framework/preemption/preemption.go:
```diff
@@ -401,6 +461,85 @@ func (ev *Evaluator) prepareCandidate(ctx context.Context, c Candidate, pod *v1.
+		ev.mu.Unlock()
+
+		if err := ev.PreemptPod(ctx, c, pod, c.Victims().Pods[len(c.Victims().Pods)-1], pluginName); err != nil {
+			logger.Error(err, "Error occurred during preemption")
```
same here

dom4ha commented on pkg/scheduler/framework/preemption/preemption.go:
```diff
@@ -401,6 +461,85 @@ func (ev *Evaluator) prepareCandidate(ctx context.Context, c Candidate, pod *v1.
+
+		ev.Handler.Parallelizer().Until(ctx, len(c.Victims().Pods)-1, preemptPod, ev.PluginName)
+		if err := errCh.ReceiveError(); err != nil {
+			logger.Error(err, "Error occurred during preemption")
```
"Error occurred during async preemption"

sanposhiho committed:
implement PodActivator to activate when preemption fails

sanposhiho commented:
/retest

sanposhiho committed:
fix: avoid errors from the race condition detector

sanposhiho commented:
/retest

dom4ha reviewed (commented):
/assign

sanposhiho committed:
fix: register Pod/Delete event at the preemption plugin

sanposhiho commented:
/retest

nayihz commented on pkg/scheduler/framework/interface.go:
```diff
@@ -898,6 +902,12 @@ func (ni *NominatingInfo) Mode() NominatingMode {
 
+// PodActivator abstracts operations in the scheduling queue.
+type PodActivator interface {
+	// Activate moves the given pods to activeQ iff they're in unschedulablePods or backoffQ.
```
```suggestion
	// Activate moves the given pods to activeQ if they're in unschedulablePods or backoffQ.
```

nayihz replied:
Ah, ignore me if you mean `if and only if` 😂

sanposhiho replied:
"if and only if", not a typo.

sanposhiho replied:
Yeah

sanposhiho replied:
Me 3 years ago did the same actually, lol 😅 
https://github.com/kubernetes/kubernetes/pull/104542

sanposhiho commented on pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go:
```diff
@@ -71,13 +73,19 @@ func New(_ context.Context, dpArgs runtime.Object, fh framework.Handle, fts feat
+		podLister: podLister,
+		pdbLister: pdbLister,
 	}
+	pl.Evaluator = preemption.NewEvaluator(Name, fh, &pl, fts.EnableAsyncPreemption)
```
`New` here is the constractor for DefaultPreemption.

sanposhiho committed:
chore: fix based on reviews

sanposhiho commented on pkg/scheduler/framework/preemption/preemption.go:
```diff
@@ -401,6 +461,91 @@ func (ev *Evaluator) prepareCandidate(ctx context.Context, c Candidate, pod *v1.
+		defer metrics.PreemptionGoroutinesExecutionTotal.WithLabelValues(result).Inc()
+		defer func() {
+			if result == metrics.GoroutineResultError {
+				// When API call isn't successful, the Pod may get stuck in the unschedulable pod pool in the worst case.
+				// So, we should move the Pod to the activeQ anyways.
+				ev.Handler.Activate(logger, map[string]*v1.Pod{pod.Name: pod})
```
Good point, I changed `Activate` to put the event in the in-flight events so that the preemptor Pods don't miss the `Activate()`.

sanposhiho committed:
fix: register the event in in-flight as necessary at Activate

sanposhiho committed:
fix: use Activate always

sanposhiho committed:
add: add a test case for Activate

sanposhiho commented:
/retest

dom4ha reviewed (commented):
/lgtm
Impressive number of tests :)

tenzen-y commented on pkg/scheduler/framework/preemption/preemption.go:
```diff
@@ -401,6 +461,71 @@ func (ev *Evaluator) prepareCandidate(ctx context.Context, c Candidate, pod *v1.
+	}
+
+	ev.mu.Lock()
+	ev.preempting[pod.UID] = struct{}{}
```
```suggestion
	ev.preempting.Insert(pod.UID)
```

nits: It seems that the `preempting` is defined as a set typed.
Is there any policy to use the map operation instead of the set operation?

https://github.com/kubernetes/kubernetes/blob/4c2ebc1e2a49e9477669bd22831fd458c32fae7f/pkg/scheduler/framework/preemption/preemption.go#L136-L138

alculquicondor replied:
Please use `Insert`

sanposhiho replied:
Thanks, 

> Is there any kube-scheduler policy using the map operation instead of the set operation?

No no, it was a map type before, someone suggested me using `set` at some point, and I just forgot changing the operation to use methods after the change.

tenzen-y commented on pkg/scheduler/framework/preemption/preemption.go:
```diff
@@ -125,10 +130,89 @@ type Evaluator struct {
+	ev.mu.RLock()
+	defer ev.mu.RUnlock()
+
+	_, ok := ev.preempting[podUID]
+	return ok
```
```suggestion
	return ev.preempting.Has(podUID)
```

nits

tenzen-y reviewed (commented):
I left a few comments for just nits.

alculquicondor commented on pkg/scheduler/backend/queue/scheduling_queue.go:
```diff
@@ -599,7 +605,11 @@ func (p *PriorityQueue) Activate(logger klog.Logger, pods map[string]*v1.Pod) {
 		}
+
+		// If this pod is in-flight, register the activation event so that the pod will be requeued when it comes back.
+		p.activeQ.addEventsIfPodInFlight(nil, pod, []framework.ClusterEvent{framework.EventForceActivate})
```
Wouldn't this event match every pod that is in-flight? Or is that desired?

Also, for future reference, can you summarize why this change is needed?

sanposhiho replied:
Oops, good catch, I forgot putting another change that must be needed on handling this activation event.

alculquicondor replied:
> can you summarize why this change is needed?

Please add this to the PR description.

sanposhiho replied:
Done

alculquicondor commented on pkg/scheduler/framework/events.go:
```diff
@@ -50,7 +53,9 @@ var (
-	EventUnschedulableTimeout = ClusterEvent{Resource: WildCard, ActionType: All, label: "UnschedulableTimeout"}
+	EventUnschedulableTimeout = ClusterEvent{Resource: WildCard, ActionType: All, label: UnschedulableTimeout}
+	// EventForceActivate is the event when a pod is moved from unschedulablePods/backoffQ to activeQ.
+	EventForceActivate = ClusterEvent{Resource: WildCard, ActionType: All, label: ForceActivate}
```
I don't think it's necessary, but I don't like the name. Do you have alternative suggestions?

sanposhiho replied:
`EventPodMoveRequest`?

alculquicondor commented on pkg/scheduler/framework/types.go:
```diff
@@ -296,7 +296,7 @@ func (ce ClusterEvent) Label() string {
 // AllClusterEventLabels returns all possible cluster event labels given to the metrics.
 func AllClusterEventLabels() []string {
-	labels := []string{EventUnschedulableTimeout.Label()}
+	labels := []string{UnschedulableTimeout, ForceActivate}
```
In general, we shouldn't be adding more special cases like this, so I don't think we need extra machinery, other than tests.

sanposhiho replied:
Yeah, this kind of special events should be a rare case. Actually, we're gonna remove `UnschedulableTimeout` at some point.

sanposhiho commented on pkg/scheduler/framework/preemption/preemption_test.go:
```diff
@@ -393,6 +412,11 @@ func TestPrepareCandidate(t *testing.T) {
 			Obj()
 
+		failVictim = st.MakePod().Name("fail-victim").UID("victim1").
+				Node(node1Name).SchedulerName(defaultSchedulerName).Priority(midPriority).
+				Containers([]v1.Container{st.MakeContainer().Name("container1").Obj()}).
+				Obj()
```
I don't think it's a semantic stuff, but gofmt formats like this.

sanposhiho committed:
fix: handle Activate event properly

sanposhiho committed:
fix: use set methods

sanposhiho committed:
fix the broken part due to the merge

sanposhiho commented:
/retest

macsko commented on pkg/scheduler/backend/queue/scheduling_queue.go:
```diff
@@ -411,6 +413,18 @@ func (p *PriorityQueue) isPodWorthRequeuing(logger klog.Logger, pInfo *framework
+		// If the wildcard event has a Pod in newObj,
+		// that indicates that the event wants to be effective for the Pod only.
+		// Specifically, EventForceActivate could have a target Pod in newObj.
+		if newObj != nil {
+			if pod, ok := newObj.(*v1.Pod); ok && pod.UID == pInfo.Pod.UID {
+				logger.V(6).Info("Worth requeuing because the event is wildcard", "pod", klog.KObj(pInfo.Pod))
+				return queueAfterBackoff
+			}
+			// This wildcard event is not for this Pod.
+			return queueSkip
+		}
```
Can it be simplified?
```suggestion
		if newObj != nil {
			if pod, ok := newObj.(*v1.Pod); !ok || pod.UID != pInfo.Pod.UID {
				// This wildcard event is not for this Pod.
			    return queueSkip
			}
		}
```

macsko commented on pkg/scheduler/backend/queue/scheduling_queue.go:
```diff
@@ -599,7 +617,13 @@ func (p *PriorityQueue) Activate(logger klog.Logger, pods map[string]*v1.Pod) {
+		// If this pod is in-flight, register the activation event so that the pod will be requeued when it comes back.
+		// Specifically in the in-tree plugins, this is for the scenario with the preemption plugin
+		// where the async preemption API calls are all done or fail at some point before the Pod comes back to the queue.
+		p.activeQ.addEventsIfPodInFlight(nil, pod, []framework.ClusterEvent{framework.EventForceActivate})
```
What if queueing hints feature is disabled and we reach this path? As inFlightEvents are not used then, pod will stay in unschedulable queue?

sanposhiho replied:
Good point! We have to update `moveRequestCycle` like we do (when QHint is disabled) every time receiving an event.

dom4ha commented on pkg/scheduler/backend/queue/scheduling_queue.go:
```diff
@@ -411,6 +413,18 @@ func (p *PriorityQueue) isPodWorthRequeuing(logger klog.Logger, pInfo *framework
+		// Specifically, EventForceActivate could have a target Pod in newObj.
+		if newObj != nil {
+			if pod, ok := newObj.(*v1.Pod); ok && pod.UID == pInfo.Pod.UID {
+				logger.V(6).Info("Worth requeuing because the event is wildcard", "pod", klog.KObj(pInfo.Pod))
```
you can add "wildcard dedicated to this pod"

sanposhiho replied:
I took macsko's suggestion and this path is no longer existing. But, I improved the log in a similar way.

sanposhiho committed:
fix: take QHint disable scenario into consideration

sanposhiho committed:
fix: improve logs\

sanposhiho commented:
/retest

sanposhiho commented:
/retest

sanposhiho committed:
fix: move to a new way updating an extended resource

sanposhiho commented:
needed to rebase on the master branch to address the change from https://github.com/kubernetes/kubernetes/pull/128194.

macsko commented on pkg/scheduler/framework/preemption/preemption_test.go:
```diff
@@ -404,34 +428,51 @@ func TestPrepareCandidate(t *testing.T) {
 		expectedStatus *framework.Status
+		// Only compared when async preemption is enabled.
+		expectedPreemptingMap sets.Set[types.UID]
+		expectedActivatedPods map[string]*v1.Pod
```
Do we need a field for this if it's always equal to map[string]*v1.Pod{preemptor.Name: preemptor}? Won't it be possible to to do this check implicitly?

macsko commented on test/integration/scheduler_perf/misc/performance-config.yaml:
```diff
@@ -354,10 +354,19 @@
+      initNodes: 5000
+      initPods: 20000
+      measurePods: 5000
+  - name: 5000Nodes_AsyncPreemptionEnabled
```
Could you set the threshold initially to 200?

sanposhiho committed:
fix: refactor TestPrepareCandidate

sanposhiho committed:
feat: set the threashold on the scheduler-perf test case

macsko commented:
/lgtm

sanposhiho commented:
/retest

alculquicondor commented on pkg/scheduler/framework/preemption/preemption.go:
```diff
@@ -401,6 +460,71 @@ func (ev *Evaluator) prepareCandidate(ctx context.Context, c Candidate, pod *v1.
+			ev.mu.Lock()
+			ev.preempting.Delete(pod.UID)
+			ev.mu.Unlock()
+			ev.Handler.Activate(logger, map[string]*v1.Pod{pod.Name: pod})
```
why are we calling Activate? IIRC, Activate was previously used by Permit plugins, right?

This seems like an overload to Activate.

IIRC, when a scheduling cycle has a failure, we reschedule the pod directly (skipping unschedulable). IIUC, we want to do roughly the same here.

sanposhiho replied:
So, the initial motivation is to address the failure case where the API call is failure at some points (https://github.com/kubernetes/kubernetes/pull/128170#discussion_r1827707578)

e.g., If the preemption API calls aren't successful at all, the preemptor Pod gets stuck in the unschedulable pod pool because no Pod/Delete event happens.
So, we have to somehow let the preemption plugin move the Pod out of the unsched pod pool in the failure case. That's why `Activate()` is a must here.

So, initially we ran `Activate()` only when the goroutine fails.
And, later we changed to always run `Activate()` after the preemption goroutine is done regardless of the goroutine result
because it can simplify the implementation and also because I _thought_ the goroutine is done == the pod is likely now schedulable.

That's all the context of the past discussion in this PR how the current `Activate()` forms.

_BUT_, rethinking about it, looks like we made a mistake. `Activate()` is a must for the failure scenario, that's correct. 
The incorrect part is the second part; the goroutine is done **!=** the pod is likely now schedulable.
The completion of the deletion APIs only means the victims just have `deletionTimestamp`, but does not mean the Pods are deleted.
So, let's say each victim takes 30 seconds (GracePeriod) to be completely deleted, the preemptor pod won't be schedulable for 30 seconds (* at least at the preemption target node).
If we `Activate()` the preemptor in that case, the preemptor's scheduling is retried and it consumes the scheduling cycle in vain.

Let me bring back the past implementation and make it run `Activate()` only when the goroutine fails

sanposhiho replied:
cc @macsko @dom4ha 
(who joins the previous discussion so far

alculquicondor replied:
> So, the initial motivation is to address the failure case where the API call is failure at some points 

This doesn't necessarily answer why we should reuse `Activate`. My pet-peeve is that this leads to adding the extra event with rare semantics and checks.

Could we just requeue the preemptor pod directly?

dom4ha replied:
Regarding the second part, my intention was to let other gating plugin to take over the gate keeping :) Apparently, there's no such functionality, although it could, as there should be still fit plugin saying no.

alculquicondor replied:
This is how we do it in `handleBindingCycleError` https://github.com/kubernetes/kubernetes/blob/62bdf5d87c1242ef5894a51f3d79e51ebcd2a07d/pkg/scheduler/schedule_one.go#L345

and this is how we do it in `handleSchedulingFailure` https://github.com/kubernetes/kubernetes/blob/62bdf5d87c1242ef5894a51f3d79e51ebcd2a07d/pkg/scheduler/schedule_one.go#L1078
Since the pod wouldn't have any failed plugins (because it was an error), it would be directly put in back off here https://github.com/kubernetes/kubernetes/blob/62bdf5d87c1242ef5894a51f3d79e51ebcd2a07d/pkg/scheduler/backend/queue/scheduling_queue.go#L705
(I'm not sure how we handled this in scheduling hints)

alculquicondor replied:
Ok, that sounds reasonable. I still don't love the approach, but there doesn't seem to be a reasonable alternative.

Please add the conclusions of this thread to the description.

sanposhiho committed:
fix: run activate() only when fail

alculquicondor commented:
In the release notes, let's be explicit about what the "expensive operations" are. It's the API calls, something that users can reasonably understand.

/approve

tenzen-y reviewed (commented):
Thanks for addressing my comments. 
LGTM from my PoV

tenzen-y commented:
BTW, it seems that we need to rebase this PR and then synchronize the API with the master branch.

dom4ha commented on pkg/scheduler/framework/preemption/preemption.go:
```diff
@@ -401,6 +460,91 @@ func (ev *Evaluator) prepareCandidate(ctx context.Context, c Candidate, pod *v1.
+		defer func() {
+			if result == metrics.GoroutineResultError {
+				// When API call isn't successful, the Pod may get stuck in the unschedulable pod pool in the worst case.
+				// So, we should move the Pod to the activeQ.
```
Note that it will be moved to the backoff queue, so it's one more argument why we should do this only in error case (I guess in the non-error case we'd rather prefer to move it to active directly).

dom4ha replied:
One more thought. Note that the removed low priority pods won't be able to jump again in this spot until the pod times out in the backoff queue. I guess it's not bad, as this place will stay reserved when the pod is retried.

I however have a different question. If some pod changes the nominated node name, is there an event that can wake up pods waiting for such a change? In our case, the removed low priority pods may remain unschedulable in case the preempted pod got eventually scheduled in different location, in which they couldn't due to other constraints. I know it's a bit theoretical and we have flush mechanism, but still.

sanposhiho replied:
> I guess in the non-error case we'd rather prefer to move it to active directly

This is a good topic to discuss. (no need to block this PR though)
In the theory that we've had so far, backoffQ is the _**penalty**_ that the pod wasted the scheduling cycle.
So, the point would be whether we should treat the scheduling cycle that has triggered the preemption as "it was wasted" or "it was necessary".

---

> I however have a different question.

Yeah, regarding the nominated node name change (= added, cleared, changed), we don't properly handle it right now. Actually, one issue is raised, see https://github.com/kubernetes/kubernetes/issues/126700#issuecomment-2347515403.

dom4ha replied:
> Yeah, regarding the nominated node name change (= added, cleared, changed), we don't properly handle it right now. Actually, one issue is raised, see [#126700 (comment)](https://github.com/kubernetes/kubernetes/issues/126700#issuecomment-2347515403). I wanted, but couldn't make time working on this nominatedNodeName problem in general.

I was thinking about this issue as well, but I missed your last update there when you confirmed it. Sounds like something relatively easy to fix.

sanposhiho replied:
If you're interested in, feel free to give it a try :)

dom4ha replied:
> If you're interested in, feel free to give it a try :)

Yeah, I can give it a try, but note that I will be OOO next week.

dom4ha commented on pkg/scheduler/backend/queue/scheduling_queue.go:
```diff
@@ -599,7 +618,14 @@ func (p *PriorityQueue) Activate(logger klog.Logger, pods map[string]*v1.Pod) {
+			continue
 		}
+
+		// If this pod is in-flight, register the activation event or update moveRequestCycle so that the pod will be requeued when it comes back.
```
"or update moveRequestCycle (when QHints are disabled)"

BTW. I was wondering how we record the events and it's actually quite clever. I'd however rename it from inFlightEvents to inFlightPodsAndEvents, as I haven't suspected that different objects are mixed.

sanposhiho replied:
Yeah, "inFlightEvents (hey, Pods are there too!)" is a bit confusing 😅

dom4ha reviewed (commented):
/lgtm

sanposhiho committed:
chore: wording

sanposhiho commented:
Yuki's replies always make me wonder whether we really are in the same zone. Done, rebased. Let's see if it resolves the openapi error.
Fixed the comment suggestion from @dom4ha too.

sanposhiho commented:
/assign @ahg-g 

For the feature gate area change approval.

tenzen-y commented:
> Yuki's replies always make me wonder whether we really are in the same timezone 😄  Done, rebased. Let's see if it resolves the openapi error.
> 
> Fixed the comment suggestion from @dom4ha too.

Haha, I'm there truly :)

sanposhiho commented:
/retest

sanposhiho commented:
/retest

sanposhiho commented:
/retest

tenzen-y commented:
@sanposhiho For unit testing, once you rebase again, those will go away, IMO

https://github.com/kubernetes/kubernetes/issues/128633

sanposhiho commented:
/retest

sanposhiho commented:
/retest

sanposhiho commented:
The alpha CI keeps failing due to unrelated tests. Please ignore, this PR is ready for review.
After @alculquicondor and @dom4ha 's approval, I only changed the comment https://github.com/kubernetes/kubernetes/pull/128170/commits/105d489aa441b1fe43b1b6db0cc98c2797998aaa.

ahg-g commented:
/approve 

for feature gate file

sanposhiho commented:
@alculquicondor @macsko @dom4ha can anyone give it a final `/lgtm`?

sanposhiho commented:
/unhold

We've got an approver's review already, although the final `/lgtm` is still necessary.

dom4ha reviewed (commented):
/lgtm
